# Proyek Pengantar Pembelajaran Mesin
## Prediksi Penyakit Liver (Hati) Menggunakan Metode KNN

### PPM D
---
Anggota :

1. Muhammad Basil Musyaffa Amin (215150201111023)
2. Muhammad Jilan Naufal (215150207111018)
3. Muhammad Fajrul Alwan (215150207111023)
4. Gibran Hakim (215150200111020)

# Preprocessing Data

### 1) Import Data

**wget** digunakan untuk mengunduh dataset dan menyimpannya dalam Google Colab. Kode pada cell di bawah ini digunakan untuk mengunduh dataset.

In [ ]:
! wget https://raw.githubusercontent.com/gevabriel/dataset/main/heart_disease.csv

--2023-05-24 10:22:02--  https://raw.githubusercontent.com/gevabriel/dataset/main/heart_disease.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10277 (10K) [text/plain]
Saving to: ‘heart_disease.csv’

heart_disease.csv   100%[===================>]  10.04K  --.-KB/s    in 0s      

2023-05-24 10:22:03 (63.3 MB/s) - ‘heart_disease.csv’ saved [10277/10277]



Setelah dataset berhasil diunduh, langkah berikutnya adalah membaca dataset dengan memanfaatkan fungsi [readcsv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) dari library pandas.

In [ ]:
import pandas as pd
import numpy as np
data = pd.read_csv('heart_disease.csv')

Tampilkan beberapa baris dari dataset untuk mendapatkan informasi singkat mengenai isi data. Gunakan fungsi **head()** untuk menampilkan 5 data pertama.

In [ ]:
data.head()

Age  Chest_pain_type   BP  Cholesterol  EKG_results  Max_HR  ST_depression  \
0   70                4  130          322            2     109            2.4   
1   67                3  115          564            2     160            1.6   
2   57                2  124          261            0     141            0.3   
3   64                4  128          263            0     105            0.2   
4   74                2  120          269            2     121            0.2   

   Slope_of_ST  Number_of_vessels_fluro  Thallium Heart_Disease  
0            2                        3         3      Presence  
1            2                        0         7       Absence  
2            1                        0         7      Presence  
3            2                        1         7       Absence  
4            1                        1         3       Absence

Berdasarkan informasi dari fungsi **head()**, data alpha thalasemia yang
digunakan mempunyai 10 fitur.

### 2) Missing value dan imputasi data

Sebelum kita melanjutkan langkah preprocessing, kita harus mengecek apakah dataset yang kita gunakan mengandung nilai dengan 0.0 atau tidak. Jika ada, kita harus mengganti nilai 0.0 tersebut dengan NaN (Not a Number). Langkah-langkah pengecakan missing value tersebut adalah sebagai berikut:


1.   Membuat filter untuk mencari data dengan N = 0, P = 0, K = 0, temperature = 0, humidity = 0, ph = 0, dan rainfall = 0
2.   Mencari data yang memenuhi kondisi1 **atau** kondisi2 **atau** kondisi3 **atau** kondisi4 **atau** kondisi5 **atau** kondisi6 **atau** kondisi7 menggunakan property [loc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html) pada dataframe



In [ ]:
kondisi1 = data['Age']==0.0
kondisi2 = data['Chest_pain_type']==0.0
kondisi3 = data['BP']==0.0
kondisi4 = data['Cholesterol']==0.0
kondisi5 = data['EKG_results']==0.0
kondisi6 = data['Max_HR']==0.0
kondisi7 = data['ST_depression']==0.0
kondisi8 = data['Slope_of_ST']==0.0
kondisi9 = data['Number_of_vessels_fluro']==0.0
kondisi10 = data['Thallium']==0.0
data.loc[kondisi1 | kondisi2 | kondisi3 | kondisi4 | kondisi5 | kondisi6 | kondisi7 | kondisi8 | kondisi9 | kondisi10]

Age  Chest_pain_type   BP  Cholesterol  EKG_results  Max_HR  \
1     67                3  115          564            2     160   
2     57                2  124          261            0     141   
3     64                4  128          263            0     105   
5     65                4  120          177            0     140   
10    59                4  135          234            0     161   
..   ...              ...  ...          ...          ...     ...   
264   48                2  110          229            0     168   
265   52                3  172          199            0     162   
266   44                2  120          263            0     173   
267   56                2  140          294            2     153   
268   57                4  140          192            0     148   

     ST_depression  Slope_of_ST  Number_of_vessels_fluro  Thallium  \
1              1.6            2                        0         7   
2              0.3            1                        0         7   
3              0.2            2                        1         7   
5              0.4            1                        0         7   
10             0.5            2                        0         7   
..             ...          ...                      ...       ...   
264            1.0            3                        0         7   
265            0.5            1                        0         7   
266            0.0            1                        0         7   
267            1.3            2                        0         3   
268            0.4            2                        0         6   

    Heart_Disease  
1         Absence  
2        Presence  
3         Absence  
5         Absence  
10        Absence  
..            ...  
264      Presence  
265       Absence  
266       Absence  
267       Absence  
268       Absence  

[220 rows x 11 columns]

Gunakan properti **replace** pada dataframe untuk menggganti 0.0 menjadi NaN

In [ ]:
data = data.replace(0.0, np.NaN)






Selanjutnya, kita perlu membuat fungsi bernama **imputasi** yang melakukan tahapan berikut:


1.   Menghitung rata-rata masing-masing kolom berdasarkan kelasnya
2.   Mengisi kolom yang berisi NaN dengan rata-rata kelas yang sesuai




In [ ]:
def imputasi(df_in):
  df_input = df_in.copy()
  list_columns = df_input.columns #mendapatkan daftar kolom pada df
  class_column = list_columns[-1] #kolom terakhir merupakan kolom kelas
  for column in list_columns[:-1]:
    df_input[column] = df_input[column].fillna(df_input.groupby(class_column)[column].transform('mean')) #penggantian nilai NaN dilakukan per kolom
  return df_input

Kami membuat sebuah dataframe baru bernama **data_imputasi** yang berisi dataset dengan nilai NaN yang sudah diganti dengan cara memanggil fungsi **imputasi**.

In [ ]:
data_imputasi = imputasi(data)

Cek apakah masih terdapat nilai NaN pada dataframe.

In [ ]:
data_imputasi.isnull().values.any()

False

Dapat dilihat bahwa pengecekan bernilai false yang berarti tidak ada yang bernilai NaN pada datasetnya.

### 3) Normalisasi MinMax
Normalisasi bertujuan menyamakan rentang nilai pada setiap fitur. Beberapa metode pembelajaran mesin memiliki kinerja yang buruk apabila rentang nilai tiap variabel berbeda jauh.

Sebelum melakukan normalisasi, cek terlebih dahulu rentang (nilai max - nilai min) pada masing masing fitur

In [ ]:
def cetak_rentang(df_input):
  list_fitur = df_input.columns[:-1]
  for fitur in list_fitur:
    max = df_input[fitur].max()
    min = df_input[fitur].min()
    print("Rentang fitur", fitur, "adalah", max-min)

In [ ]:
cetak_rentang(data_imputasi)

Rentang fitur Age adalah 48
Rentang fitur Chest_pain_type adalah 3
Rentang fitur BP adalah 106
Rentang fitur Cholesterol adalah 438
Rentang fitur EKG_results adalah 1.0
Rentang fitur Max_HR adalah 131
Rentang fitur ST_depression adalah 6.1000000000000005
Rentang fitur Slope_of_ST adalah 2
Rentang fitur Number_of_vessels_fluro adalah 2.0
Rentang fitur Thallium adalah 4


Terlihat bahwa masing-masing fitur memiliki rentang yang berbeda, meskipun tidak terlalu signifikan. Cara pertama yang bisa dilakukan adalah dengan metode minmax, dengan tujuan mengubah skala setiap fitur ke interval [0,1] dengan cara menggeser nilai setiap fitur sehingga nilai minimal adalah 0, dan kemudian membagi dengan nilai maksimal baru.


In [ ]:
def minmax(df_input):
  list_fitur = df_input.columns[:-1]
  for fitur in list_fitur:
    max = df_input[fitur].max()
    min = df_input[fitur].min()
    df_input[fitur] = (df_input[fitur]-min)/(max-min)
  return df_input

Kami membuat sebuah dataframe baru bernama **data_normal** yang berisi hasil dari metode minmax dengan input **data_imputasi**

In [ ]:
data_normal = minmax(data_imputasi)

Cek 5 baris pertama **data_normal**

In [ ]:
data_normal.head()

Age  Chest_pain_type        BP  Cholesterol  EKG_results    Max_HR  \
0  0.854167         1.000000  0.339623     0.447489     1.000000  0.290076   
1  0.791667         0.666667  0.198113     1.000000     1.000000  0.679389   
2  0.583333         0.333333  0.283019     0.308219     0.986486  0.534351   
3  0.729167         1.000000  0.320755     0.312785     0.984615  0.259542   
4  0.937500         0.333333  0.245283     0.326484     1.000000  0.381679   

   ST_depression  Slope_of_ST  Number_of_vessels_fluro  Thallium Heart_Disease  
0       0.377049          0.5                 1.000000       0.0      Presence  
1       0.245902          0.5                 0.216667       1.0       Absence  
2       0.032787          0.0                 0.362500       1.0      Presence  
3       0.016393          0.5                 0.000000       1.0       Absence  
4       0.016393          0.0                 0.000000       0.0       Absence

Tampilkan rentang masing-masing fitur menggunakan fungsi **cetak_rentang** yang telah dibuat

In [ ]:
cetak_rentang(data_normal)

Rentang fitur Age adalah 1.0
Rentang fitur Chest_pain_type adalah 1.0
Rentang fitur BP adalah 1.0
Rentang fitur Cholesterol adalah 1.0
Rentang fitur EKG_results adalah 1.0
Rentang fitur Max_HR adalah 1.0
Rentang fitur ST_depression adalah 1.0
Rentang fitur Slope_of_ST adalah 1.0
Rentang fitur Number_of_vessels_fluro adalah 1.0
Rentang fitur Thallium adalah 1.0


Dapat dilihat bahwa interval rentang dari setiap fitur sudah diantara 0 hingga 1.

## 4) Normalisasi Z-Score
Cara lain yang bisa digunakan dalam normalisasi data adalah dengan Z-score, dengan tujuan hasil dari normalisasi tersebut adalah mendapatkan rata-rata dengan skor 0 dan standar deviasi 1


In [ ]:
def zscore(df_in):
  df_input = df_in.copy()
  list_fitur = df_input.columns[:-1]
  for fitur in list_fitur:
    mean = df_input[fitur].mean()
    std = df_input[fitur].std()
    df_input[fitur] = (df_input[fitur]-mean)/(std)
  return df_input

In [ ]:
data_zscore = zscore(data_imputasi)

In [ ]:
data_zscore.head()

Age  Chest_pain_type        BP  Cholesterol  EKG_results    Max_HR  \
0  1.708920         0.869313 -0.075270     1.399613     0.170000 -1.755947   
1  1.379578        -0.183219 -0.915060     6.081711     0.170000  0.445582   
2  0.281771        -1.235750 -0.411186     0.219415     0.012143 -0.374596   
3  1.050236         0.869313 -0.187242     0.258110    -0.009714 -1.928616   
4  2.148043        -1.235750 -0.635130     0.374195     0.170000 -1.237940   

   ST_depression  Slope_of_ST  Number_of_vessels_fluro  Thallium Heart_Disease  
0       1.014905     0.675165                 2.879690 -0.874083      Presence  
1       0.152481     0.675165                -0.259766  1.187073       Absence  
2      -1.248957    -0.952466                 0.324707  1.187073      Presence  
3      -1.356760     0.675165                -1.128126  1.187073       Absence  
4      -1.356760    -0.952466                -1.128126 -0.874083       Absence

In [ ]:
print('Cek Nilai Rentang:')
print('Sebelum Normalisasi')
cetak_rentang(data_imputasi)
print('\nNormalisasi MinMax')
cetak_rentang(data_normal)
print('\nNormalisasi ZScore')
cetak_rentang(data_zscore)

print('\nCek Rata-Rata dan Standar Deviasi:')
print('Sebelum Normalisasi')
print(round(data_imputasi.describe(),3))
print('\nNormalisasi MinMax')
print(round(data_normal.describe(),3))
print('\nNormalisasi ZScore')
print(round(data_zscore.describe(),3))

Cek Nilai Rentang:
Sebelum Normalisasi
Rentang fitur Age adalah 1.0
Rentang fitur Chest_pain_type adalah 1.0
Rentang fitur BP adalah 1.0
Rentang fitur Cholesterol adalah 1.0
Rentang fitur EKG_results adalah 1.0
Rentang fitur Max_HR adalah 1.0
Rentang fitur ST_depression adalah 1.0
Rentang fitur Slope_of_ST adalah 1.0
Rentang fitur Number_of_vessels_fluro adalah 1.0
Rentang fitur Thallium adalah 1.0

Normalisasi MinMax
Rentang fitur Age adalah 1.0
Rentang fitur Chest_pain_type adalah 1.0
Rentang fitur BP adalah 1.0
Rentang fitur Cholesterol adalah 1.0
Rentang fitur EKG_results adalah 1.0
Rentang fitur Max_HR adalah 1.0
Rentang fitur ST_depression adalah 1.0
Rentang fitur Slope_of_ST adalah 1.0
Rentang fitur Number_of_vessels_fluro adalah 1.0
Rentang fitur Thallium adalah 1.0

Normalisasi ZScore
Rentang fitur Age adalah 5.269475184319818
Rentang fitur Chest_pain_type adalah 3.1575954834650592
Rentang fitur BP adalah 5.934515988838667
Rentang fitur Cholesterol adalah 8.474209469212031
Ren

Dapat dilihat pada tabel normalisasi Z-Score, didapatkan rata-rata (mean) mendekati 0 dan standar deviasi (std) bernilai 1.

# Klasifikasi KNN

### 1) Membagi data menjadi data latih dan data uji

Metode pembelajaran mesin memerlukan dua jenis data :


1.   Data latih : Digunakan untuk proses training metode klasifikasi
2.   Data uji : Digunakan untuk proses evaluasi metode klasifikasi

Data uji dan data latih perlu dibuat terpisah (mutualy exclusive) agar hasil evaluasi lebih akurat.

Data uji dan data latih dapat dibuat dengan cara membagi dataset dengan rasio tertentu, misalnya 80% data latih dan 20% data uji.

Library Scikit-learn memiliki fungsi [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) pada modul **model_selection** untuk membagi dataset menjadi data latih dan data uji. Bagilah dataset anda menjadi dua, yaitu **data_latih** dan **data_uji**.


In [ ]:
from sklearn.model_selection import train_test_split
data_latih, data_uji = train_test_split(data_imputasi, test_size = 0.2)

Tampilkan banyaknya data pada **data_latih** dan **data_uji**. Seharusnya **data_latih** terdiri dari 120 data, dan **data_uji** terdiri dari 30 data

In [ ]:
print(data_latih.shape[0])
print(data_uji.shape[0])

216
54


Pisahkan label/kategori dari data latih dan data uji menjadi variabel tersendiri. Beri nama variabelnya **label_latih** dan **label_uji**

In [ ]:
label_latih = data_latih.pop('Heart_Disease')
label_uji = data_uji.pop('Heart_Disease')

In [ ]:
data_uji.head()

Age  Chest_pain_type        BP  Cholesterol  EKG_results    Max_HR  \
113  0.520833         0.666667  0.386792     0.406393     0.984615  0.755725   
19   0.229167         0.000000  0.433962     0.166667     0.984615  0.816794   
35   0.395833         1.000000  0.339623     0.296804     1.000000  0.603053   
252  0.312500         1.000000  0.169811     0.374429     1.000000  0.625954   
99   0.437500         0.333333  0.245283     0.269406     0.984615  0.694656   

     ST_depression  Slope_of_ST  Number_of_vessels_fluro  Thallium  
113       0.159601          0.0                 0.216667       0.0  
19        0.213115          0.0                 0.216667       1.0  
35        0.301606          0.0                 0.500000       1.0  
252       0.301606          0.0                 0.000000       0.0  
99        0.163934          0.0                 0.216667       0.0

## 3) Menghitung jarak euclidean

Tahapan awal dari algoritma KNN adalah perhitungan jarak. Salah satu metode perhitungan jarak yang bisa digunakan adalah jarak Euclidean. Buatlah fungsi bernama **jarakEu** yang berfungsi menghitung jarak euclidean dari dua buah vektor (tanpa kelas)

In [ ]:
def jarakEu(data1, data2):
  jarak = np.square(data1 - data2)
  jarak = np.sum(jarak)
  return np.sqrt(jarak)

Uji fungsi **jarakEu** untuk menghitung jarak antara data latih pertama dengan data uji pertama.

In [ ]:
jarak = jarakEu(data_latih.iloc[0], data_uji.iloc[0])
print(jarak)

1.1151968863512662


## 4) Algoritma KNN

Implementasikan algoritma KNN dengan tahapan-tahapan berikut :
1. Hitung jarak antara data uji dengan setiap data latih.
2. Cari *k* data latih dengan jarak terkecil.
3. Tentukan kelas dari *k* data latih tersebut.
4. Kelas data uji ditentukan dari mayoritas kelas *k* data latih.

Buatlah fungsi dengan nama **knn** yang menerima input berupa *k*, sekumpulan data latih dan labelnya, serta sebuah data uji

In [ ]:
from collections import Counter
def knn(k, datalatih, labellatih, datauji):
  jarak = np.array([jarakEu(datalatih.iloc[x],datauji) for x in range(datalatih.shape[0])])
  indeks_k_minimum = jarak.argsort()[:k]
  k_kelas = labellatih.iloc[indeks_k_minimum].to_numpy()
  counter = Counter(k_kelas)
  kelas_uji = counter.most_common(1)[0][0]
  return kelas_uji

Lakukan pengujian fungsi **knn** untuk menentukan kelas dari data uji pertama

In [ ]:
hasil = knn(3,data_latih,label_latih,data_uji.iloc[0])
print(hasil)

Absence


## 5) KNN pada Scikit-learn

Library scikit-learn memiliki algoritma KNN yang siap untuk digunakan. Sebagai pengenalan, Anda akan melakukan klasifikasi data yang sama menggunakan KNN yang tersedia pada library scikit-learn

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
KNN = KNeighborsClassifier(n_neighbors = 3)
KNN.fit(data_latih, label_latih)
kelas = KNN.predict([data_uji.iloc[0]])
print(kelas)

['Absence']


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


# Evaluasi Klasifikasi

## 1) Tentukan metode klasifikasi yang akan dievaluasi

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

models = []
models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC(gamma='auto')))

In [ ]:
# evaluate each model in turn
results = []
names = []

## 2) Baca dataset


In [ ]:
data_imputasi

Age  Chest_pain_type        BP  Cholesterol  EKG_results    Max_HR  \
0    0.854167         1.000000  0.339623     0.447489     1.000000  0.290076   
1    0.791667         0.666667  0.198113     1.000000     1.000000  0.679389   
2    0.583333         0.333333  0.283019     0.308219     0.986486  0.534351   
3    0.729167         1.000000  0.320755     0.312785     0.984615  0.259542   
4    0.937500         0.333333  0.245283     0.326484     1.000000  0.381679   
..        ...              ...       ...          ...          ...       ...   
265  0.479167         0.666667  0.735849     0.166667     0.984615  0.694656   
266  0.312500         0.333333  0.245283     0.312785     0.984615  0.778626   
267  0.562500         0.333333  0.433962     0.383562     1.000000  0.625954   
268  0.583333         1.000000  0.433962     0.150685     0.984615  0.587786   
269  0.791667         1.000000  0.622642     0.365297     1.000000  0.282443   

     ST_depression  Slope_of_ST  Number_of_vessels_fluro  Thallium  \
0         0.377049          0.5                 1.000000      0.00   
1         0.245902          0.5                 0.216667      1.00   
2         0.032787          0.0                 0.362500      1.00   
3         0.016393          0.5                 0.000000      1.00   
4         0.016393          0.0                 0.000000      0.00   
..             ...          ...                      ...       ...   
265       0.065574          0.0                 0.216667      1.00   
266       0.159601          0.0                 0.216667      1.00   
267       0.196721          0.5                 0.216667      0.00   
268       0.049180          0.5                 0.216667      0.75   
269       0.229508          0.5                 1.000000      0.00   

    Heart_Disease  
0        Presence  
1         Absence  
2        Presence  
3         Absence  
4         Absence  
..            ...  
265       Absence  
266       Absence  
267       Absence  
268       Absence  
269      Presence  

[270 rows x 11 columns]

## 3) Membuat Set Data Validasi
Tujuan evaluasi adalah memperkirakan keakuratan model terbaik pada data tak terlihat dengan mengevaluasinya pada data tak terlihat aktual. Evaluasi ini dilakukan dengan cara membagi dataset menjadi dua bagian yakni sebagian digunakan sebagai data latih dan sebagian yang lain digunakan sebagai data uji. Beberapa praktikum sebelumnya telah dipraktekkan dengan menggunakan metode train_test_split :
X_train, X_validation, Y_train, Y_validation = train_test_split(X, y, test_size=0.20, random_state=1)


Sebelum membuat data latih dan data uji. Langkah pertama pisahkan antara data attribute dan data label suatu dataset. Karena dataset liver_patient terdiri dari 9 attribut dan 1 label.  Maka kita catat X sebagai data dan Y sebagai label.

In [ ]:
from sklearn.model_selection import train_test_split
array = data_imputasi.values
X = array[:,0:10]
y = array[:,10]
X_train, X_validation, Y_train, Y_validation = train_test_split(X, y, test_size=0.20, random_state=1, shuffle=True)

## 4) K-Fold Cross Validation
k-Fold Cross-Validation Validasi silang adalah prosedur pengambilan sampel ulang yang digunakan untuk mengevaluasi model pembelajaran mesin pada sampel data terbatas. Prosedur ini memiliki parameter tunggal yang disebut k yang mengacu pada jumlah grup yang akan dipecah menjadi sampel data tertentu. Karena itu, prosedur ini sering disebut k-fold cross-validation. Ketika nilai spesifik untuk k dipilih, nilai tersebut dapat digunakan sebagai pengganti k dalam referensi ke model, seperti k = 10 menjadi validasi silang 10 kali lipat.

Validasi silang terutama digunakan dalam pembelajaran mesin terapan untuk memperkirakan kemampuan model pembelajaran mesin pada data yang tidak terlihat. Yaitu, menggunakan sampel terbatas untuk memperkirakan bagaimana performa model secara umum saat digunakan untuk membuat prediksi pada data yang tidak digunakan selama pelatihan model.

Prosedur umumnya adalah sebagai berikut:
- Acak kumpulan data berdasarkan id
- Pisahkan kumpulan data menjadi kelompok k
- Untuk setiap grup unik:
  - Ambil kelompok sebagai kelompok data uji
  - Ambil kelompok yang tersisa sebagai kumpulan data pelatihan
  - Catat hasil pengujian masing-masing group

K-Fold ini digunakan untuk validasi model yang digunakan. Evaluasi dilakukan pada dataset latih. Pada praktikum ini digunakan k=10, berarti akan membagi dataset latih menjadi 10 bagian. Satu bagian digunakan untuk testing dan sisanya sebagai data latih.

In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import StratifiedKFold

for name, model in models:
  kfold = StratifiedKFold(n_splits=10, random_state=1, shuffle=True)
  cv_results = cross_val_score(model, X_train, Y_train, cv=kfold, scoring='accuracy')
  results.append(cv_results)
  names.append(name)
  print('%s: %f (%f)' % (name, cv_results.mean(), cv_results.std()))

KNN: 0.810390 (0.077586)
NB: 0.782468 (0.078104)
SVM: 0.791775 (0.080593)


## 5) Membuat Prediksi
Kita dapat menyesuaikan model pada seluruh dataset pelatihan dan membuat prediksi pada dataset validasi.

In [ ]:
# metode knn
model = KNeighborsClassifier()
model.fit(X_train, Y_train)
predictions = model.predict(X_validation)

## 6) Evaluasi Prediksi
Kami dapat mengevaluasi prediksi dengan membandingkannya dengan hasil yang diharapkan dalam set validasi, lalu menghitung akurasi klasifikasi, serta matriks kebingungan dan laporan klasifikasi.

In [ ]:
from sklearn.datasets import make_classification
X, y = make_classification(n_samples=1000, n_features=20, n_informative=15, n_redundant=5, random_state=1)

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(Y_validation, predictions))

print(Y_validation)
print()
print(predictions)

              precision    recall  f1-score   support

     Absence       0.78      0.81      0.79        31
    Presence       0.73      0.70      0.71        23

    accuracy                           0.76        54
   macro avg       0.75      0.75      0.75        54
weighted avg       0.76      0.76      0.76        54

['Absence' 'Presence' 'Absence' 'Presence' 'Presence' 'Presence' 'Absence'
 'Presence' 'Absence' 'Presence' 'Presence' 'Presence' 'Presence'
 'Absence' 'Absence' 'Presence' 'Absence' 'Presence' 'Presence' 'Absence'
 'Presence' 'Absence' 'Absence' 'Absence' 'Absence' 'Presence' 'Absence'
 'Absence' 'Absence' 'Absence' 'Presence' 'Absence' 'Absence' 'Presence'
 'Absence' 'Absence' 'Absence' 'Absence' 'Absence' 'Presence' 'Absence'
 'Absence' 'Presence' 'Absence' 'Absence' 'Absence' 'Absence' 'Presence'
 'Presence' 'Presence' 'Absence' 'Presence' 'Absence' 'Presence']

['Absence' 'Presence' 'Absence' 'Presence' 'Presence' 'Absence' 'Presence'
 'Absence' 'Absence' 'Pre